packages/flutter_tools/lib/src/ios/ios_project_migration.dart

```diff
@@ -0,0 +1,144 @@
+        _xcodeProjectInfoFile = project.xcodeProjectInfoFile,
+        _podfile = project.podfile,
+        _logger = logger;
+
+  final File _xcodeProjectInfoFile;
+  final File _podfile;
+  final Logger _logger;
+
+  /// Inspect [project] for necessary migrations and rewrite files as needed.
+  void migrate() {
+    _migrateXcodeProjectInfoFile();
+    _migratePodfile();
+  }
+
+  void _migrateXcodeProjectInfoFile() {
+    if (!_xcodeProjectInfoFile.existsSync()) {
+      _logger.printTrace('Xcode project not found, skipping migration');
+      return;
+    }
+
+    _processFileLines(_xcodeProjectInfoFile, (String line) {
+      // App.framework Frameworks reference.
+      // isa = PBXFrameworksBuildPhase;
+      // files = (
+      //    3B80C3941E831B6300D905FE /* App.framework in Frameworks */,
+      if (line.contains('3B80C3941E831B6300D905FE')) {
+        return null;
+      }
+
+      // App.framework Embed Framework reference (build phase to embed framework).
+      // 3B80C3951E831B6300D905FE /* App.framework in Embed Frameworks */,
+      if (line.contains('3B80C3951E831B6300D905FE')
+          || line.contains('741F496821356857001E2961')) { // Ephemeral add-to-app variant.
+        return null;
+      }
+
+      // App.framework project file reference (seen in Xcode navigator pane).
+      // isa = PBXGroup;
+      // children = (
+      //	 3B80C3931E831B6300D905FE /* App.framework */,
+      if (line.contains('3B80C3931E831B6300D905FE')
+          || line.contains('741F496521356807001E2961')) { // Ephemeral add-to-app variant.
+        return null;
+      }
+
+      // Flutter.framework Frameworks reference.
+      // isa = PBXFrameworksBuildPhase;
+      // files = (
+      //   9705A1C61CF904A100538489 /* Flutter.framework in Frameworks */,
+      if (line.contains('9705A1C61CF904A100538489')) {
+        return null;
+      }
+
+      // Flutter.framework Embed Framework reference (build phase to embed framework).
+      // 9705A1C71CF904A300538489 /* Flutter.framework in Embed Frameworks */,
+      if (line.contains('9705A1C71CF904A300538489')
+          || line.contains('741F496221355F47001E2961')) { // Ephemeral add-to-app variant.
+        return null;
+      }
+
+      // Flutter.framework project file reference (seen in Xcode navigator pane).
+      // isa = PBXGroup;
+      // children = (
+      //	 9740EEBA1CF902C7004384FC /* Flutter.framework */,
+      if (line.contains('9740EEBA1CF902C7004384FC')
+          || line.contains('741F495E21355F27001E2961')) { // Ephemeral add-to-app variant.
+        return null;
+      }
+
+      // Embed and thin frameworks in a script instead of using Xcode's link / embed build phases.
+      const String thinBinaryScript = 'xcode_backend.sh\\" thin';
+      if (line.contains(thinBinaryScript)) {
+        return line.replaceFirst(thinBinaryScript, 'xcode_backend.sh\\" embed_and_thin');
+      }
+
+      return line;
+    });
+  }
+
+  void _migratePodfile() {
```

Comment:
Breadcrumb for later reference: I think we can use this to fix https://github.com/flutter/flutter/issues/39659

Reply:
I would love to move most of that logic into a Ruby helper script in the project (like xcode_backend) so it never has to be upgraded again https://github.com/flutter/flutter/issues/45197